File: dotnet/src/Pactman.NonprofitCheckPlus/Models/Pub78Source.cs
using System.Collections.Generic;
using System.Text.Json;

namespace Pactman.NonprofitCheckPlus.Models
{
    /// <summary>IRS Publication 78 findings.</summary>
    public sealed class Pub78Source : SourceView
    {
        /// <summary>Initializes the view over the projected fields.</summary>
        /// <param name="fields">The fields copied from the organization.</param>
        public Pub78Source(IEnumerable<KeyValuePair<string, JsonElement>> fields)
            : base(fields)
        {
        }

        /// <summary>Whether the organization appears in Publication 78.</summary>
        public bool? Verified => GetBoolean("verified");

        /// <summary>The organization's name as Publication 78 records it.</summary>
        public string? OrganizationName => GetString("organization_name");

        /// <summary>The EIN as Publication 78 records it.</summary>
        public string? Ein => GetString("ein");

        /// <summary>The city as Publication 78 records it.</summary>
        public string? City => GetString("city");

        /// <summary>The state as Publication 78 records it.</summary>
        public string? State => GetString("state");

        /// <summary>The Publication 78 deductibility indicator.</summary>
        public string? Indicator => GetString("indicator");

        /// <summary>The note for organizations the IRS treats as churches.</summary>
        public string? ChurchMessage => GetString("church_message");

        /// <summary>
        /// The Publication 78 deductibility entries, as the API sent them.
        /// </summary>
        /// <remarks>
        /// Empty when the API sent none, so it is always safe to enumerate; ask
        /// <c>Has("organization_types")</c> when the difference matters.
        /// </remarks>
        public IReadOnlyList<OrganizationType> OrganizationTypes =>
            OrganizationType.ListFrom(GetElement("organization_types"));

        /// <summary>The date of the most recent Publication 78 extract behind this record.</summary>
        public string? MostRecent => GetString("most_recent");
    }
}

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/models/Pub78Source.java
package org.pactman.nonprofitcheckplus.models;

import java.util.List;
import java.util.Map;

/** IRS Publication 78 findings. */
public final class Pub78Source extends SourceView {

    /**
     * Initializes the view over the projected fields.
     *
     * @param fields the fields copied from the organization.
     */
    public Pub78Source(Map<String, Object> fields) {
        super(fields);
    }

    /**
     * Whether the organization appears in Publication 78.
     *
     * @return the flag, or {@code null} when the API returned none.
     */
    public Boolean getVerified() {
        return getBoolean("verified");
    }

    /**
     * The organization's name as Publication 78 records it.
     *
     * @return the name, or {@code null} when the API returned none.
     */
    public String getOrganizationName() {
        return getString("organization_name");
    }

    /**
     * The EIN as Publication 78 records it.
     *
     * @return the EIN, or {@code null} when the API returned none.
     */
    public String getEin() {
        return getString("ein");
    }

    /**
     * The city as Publication 78 records it.
     *
     * @return the city, or {@code null} when the API returned none.
     */
    public String getCity() {
        return getString("city");
    }

    /**
     * The state as Publication 78 records it.
     *
     * @return the state, or {@code null} when the API returned none.
     */
    public String getState() {
        return getString("state");
    }

    /**
     * The Publication 78 deductibility indicator.
     *
     * @return the indicator, or {@code null} when the API returned none.
     */
    public String getIndicator() {
        return getString("indicator");
    }

    /**
     * The note for organizations the IRS treats as churches.
     *
     * @return the note, or {@code null} when the API returned none.
     */
    public String getChurchMessage() {
        return getString("church_message");
    }

    /**
     * The Publication 78 deductibility entries, as the API sent them.
     *
     * <p>Empty when the API sent none, so it is always safe to iterate; ask
     * {@code has("organization_types")} when the difference matters.
     *
     * @return the entries.
     */
    public List<OrganizationType> getOrganizationTypes() {
        return OrganizationType.listFrom(get("organization_types"));
    }

    /**
     * The date of the most recent Publication 78 extract behind this record.
     *
     * @return the date as the API formats it, or {@code null}.
     */
    public String getMostRecent() {
        return getString("most_recent");
    }
}
